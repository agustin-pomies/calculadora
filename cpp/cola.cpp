#include "../include/cola.hpp"
#include <stddef.h>

struct rep_nodo {
	int dato;
	rep_nodo* sig;
};

struct rep_cola {
	rep_nodo* ppio;
	rep_nodo* fin;
};

cola crear_cola() {
	cola nuevo = new rep_cola;
	nuevo->ppio = nuevo->fin = NULL;
	return nuevo;
}

void encolar(int b, cola &c) {
	rep_nodo* elem = new rep_nodo;
	elem->dato = b;
	elem->sig = NULL;
	if(es_vacia_cola(c))
		c->ppio = elem;
	else
		c->fin->sig = elem;
	c->fin = elem;
}

void desencolar(cola &c) {
	if(!es_vacia_cola(c)) {
		rep_nodo* aux = c->ppio;
		if(c->ppio->sig == NULL)
			c->fin = NULL;
		c->ppio = c->ppio->sig;
		delete(aux);
	};
}

bool es_vacia_cola(cola c) {
	return c->ppio == NULL;
}

int frente(cola c) {
	return c->ppio->dato;
}

void liberar_cola(cola &c) {
	while(!es_vacia_cola(c))
		desencolar(c);
	delete(c);
}
