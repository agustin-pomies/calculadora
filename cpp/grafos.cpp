#include "../include/grafos.hpp"

// GRAFOS
// Metodo para ingresar grafos:
//      & Ingresando el nro de vertices y las aristas del grafo
//      & Ingresando la matriz de incidencia
//      & Ingresando por listas de vecinos
