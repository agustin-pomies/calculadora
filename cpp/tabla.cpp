#include "../include/tabla.hpp"
#include "../include/lista.hpp"
#include <stdlib.h>
#include <limits.h>

struct rep_tabla {
  nodo *elems;
  int tamanio;
};

int hash(char c, tabla t) { return (int)c % t->tamanio; } 

tabla crear_tabla(int tamanio) {
	tabla t = new rep_tabla;
	t->tamanio = tamanio;
	t->elems = new nodo [tamanio];
	for(int i = 0; i < tamanio; i++)
		t->elems[i] = NULL;
	return t;
}

void insertar_en_tabla(char clave, int valor, tabla &t) {
	int indice = hash(clave, t);
	insertar_en_lista(clave, valor, t->elems[indice]);
}

void eliminar_de_tabla(char clave, tabla &t) {
	int indice = hash(clave, t);
	eliminar_de_tabla(clave, t->elems[indice]);
}

void liberar_tabla(tabla &t) {
	for(int i = 0; i < t->tamanio; i++)
		liberar_lista(t->elems[i]);
	delete t->elems;
	delete t;
}

int valor_en_tabla(char clave, tabla t) {
	int indice = hash(clave, t);
	int n = valor_en_lista(clave, t->elems[indice]);
	return n;
}

tabla copiar_tabla(tabla t) {
	tabla copia = crear_tabla(t->tamanio);
		for(int i = 0; i < t->tamanio; i++)
			copia->elems[i] = copiar_lista(t->elems[i]);
	return copia;
}