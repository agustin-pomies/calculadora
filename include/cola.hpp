#ifndef _COLA_H
#define _COLA_H

struct rep_cola;
typedef rep_cola *cola;

/* Constructoras */

/*  Devuelve una cola_binarios vacía (sin elementos). */
cola crear_cola();

/* Encola `a' en `c'. */
void encolar(char a, cola &c);

/* Destructoras */

/*
  Remueve de `c' el elemento que está en el frente.
  No libera la memoria del elemento removido.
  Si es_vacia_cola(c) no hace nada.
 */
void desencolar(cola &c);

/* Libera la memoria asignada a `c', pero NO la de sus elementos. */
void liberar_cola(cola &c);

/* Predicados */

/* Devuelve `true' si y sólo si `c' es vacía (no tiene elementos). */
bool es_vacia_cola(cola c);

/* Selectoras */

/*
  Devuelve el elemento que está en el frente de `c'.
  Precondición: ! es_vacia_cola(c);
 */
char frente(cola c);

#endif
