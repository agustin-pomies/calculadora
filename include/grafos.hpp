#ifndef _GRAFOS_H
#define _GRAFOS_H

//#include "matrices.hpp"
//#include "cola.hpp"

// GRAFOS
// Metodo para ingresar grafos:
//      & Ingresando el nro de vertices y las aristas del grafo (ad, bc, ...) [Lista de incidencia]
//      & Ingresando la matriz de incidencia [Matriz de adyacencia]
//      & Ingresando por listas de vecinos [Lista de adyacencia]

typedef int num_t;
typedef int *grafo;
typedef int *vertice;

// Sea n la cantidad de vertices, crea una matriz de n*n
// Precondicion: n > 0
grafo crear_grafo(num_t n);

// Remueve del grafo el vertice a
// Si ya estaba ese vertice no hace nada
void agregar_vertice(grafo &G, vertice a);

// Remueve del grafo el vertice a y sus aristas
// Si no estaba ese vertice no hace nada
void remover_arista(grafo &G, vertice a);

// Agrega al grafo la arista entre a y b
// Si la arista ya existia no hace nada
// Precondicion: es_vertice(a) & es_vertice(b)
void agregar_arista(grafo &G, vertice a, vertice b);

// Remueve del grafo la arista entre a y b
// Si no habia arista no hace nada
// Precondicion: es_vertice(a) & es_vertice(b)
void remover_arista(grafo &G, vertice a, vertice b);

// Devuelve el nro de vertices de G
num_t nro_vertices(grafo G);

// Devuelve el nro de aristas de G
num_t nro_aristas(grafo G);

// Devuelve el grado del vertice v en el grafo G
// Precondicion: es_vertice(v)
num_t grado(grafo G, vertice v);

// Devuelve la distancia entre los vertices a y b
// Precondicion: es_vertice(a) & es_vertice(b)
num_t distancia(grafo G, vertice a, vertice b);

// Devuelve la mayor distancia entre dos vertices cualesquiera del grafo G
num_t diametro(grafo G);

// Devuelve el nro de caminos de largo n que comienzan en a y terminan en b
// Precondicion: n > 0, es_vertice(a), es_vertice(b)
num_t nro_caminos(grafo G, vertice a, vertice b, num_t n);

// Devuelve true si v es vertice de G, false en caso contrario
bool es_vertice(grafo G, vertice v);

// Devuelve true si entre a y b hay una arista que los une, false en caso contrario
bool son_adyacentes(vertice a, vertice b);

// Devuelve true si G esta vacio false en caso contrario
bool es_vacio(grafo G);

// Devuelve true si G es completo, false en caso contrario
bool es_completo(grafo G);

// Devuelve una copia del grafo G
grafo copiar_grafo(grafo G);

// Devuelve el grafo complementario de G
grafo complementario(grafo G);

// Devuelve true si G es conexo, false en caso contrario
bool es_conexo(grafo G);

// Devuelve el nro de componentes conexas de G
num_t componentes_conexas(grafo G);

// Devuelve true si G es plano, false en caso contrario
bool es_plano(grafo G);

// Devuelve true si G es aciclico, false en caso contrario
bool es_aciclico(grafo G);

// Devuelve true si G es arbol, false en caso contrario
bool es_arbol(grafo G);

// Devuelve true si el G tiene un recorrido euleriano, false en caso contrario
bool existe_rec_euleriano(grafo G);

// Devuelve true si el G tiene un circuito euleriano, false en caso contrario
bool existe_circ_euleriano(grafo G);

// Devuelve true si el G tiene un camino hamiltoniano, false en caso contrario
bool existe_cam_hamiltoniano(grafo G);

// Devuelve true si el G tiene un ciclo hamiltoniano, false en caso contrario
bool existe_cic_hamiltoniano(grafo G);

// COLORACION

#endif