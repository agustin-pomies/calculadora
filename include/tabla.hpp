struct rep_tabla;

typedef rep_tabla *tabla;

/*
 Crea una tabla de asociaciones texto->numero.
 Podrá haber hasta `tamanio' asociaciones.
 */
tabla crear_tabla(int tamanio);

/*
  Inserta en `t' la asociación entre `clave' y `valor'.
  Si ya existía una asociación para `clave' es reemplazada por la
  nueva.
  Precondición: valor != INT_MAX.
  El tiempo de ejecución es O(1).
 */
void insertar_en_tabla(char clave, int valor, tabla &t);

/*
  Elimina de `t' la asociación entre `clave' y algún valor y libera la menoria
  asignada a mantener esa asociación.
  El tiempo de ejecución es O(1) en promedio.
 */
void eliminar_de_tabla(char clave, tabla &t);

/*
 Libera la memoria asignada a `t' y todos sus elementos.
 */
void liberar_tabla(tabla &t);

/*
  Devuelve el valor correspondiente a la asociación de `clave' en `t', o INT_MAX
  si no existe tal asociación.
  El tiempo de ejecución es O(1) en promedio.
 */
int valor_en_tabla(char clave, tabla t);
