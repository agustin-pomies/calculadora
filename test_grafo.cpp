#include "include/grafos.hpp"
#include "include/tabla.hpp"
#include "include/cola.hpp"

#include <stdio.h>

/* LISTA DE FUNCIONES A PROBAR

bool esta_lleno_grafo(grafo G);
grafo copiar_grafo(grafo G);
grafo complementario(grafo G);
*/

const char et1 = 'A';
const char et2 = 'B';
const char et3 = 'C';
const char et4 = 'D';
const char et5 = 'E';
const char et6 = 'F';
const char et7 = 'G';
const char et8 = 'H';
const char et9 = 'I';
const char et10 = 'J';

int main() {
    // GRAFO VACIO
    printf("PRUEBA CON GRAFO VACIO\n");
    grafo vacio = crear_grafo();
    if(es_grafo_vacio(vacio))
        printf("El grafo es vacio.\n");
    else
        printf("El grafo es NO vacio.\n");
        
    if(es_vertice(vacio, et1))
        printf("El grafo TIENE el vertice %c.\n", et1);
    else
        printf("El grafo NO TIENE el vertice %c.\n", et1);
    eliminar_grafo(vacio);
    printf("\n");
    
    // GRAFO K1
    printf("PRUEBA CON K1\n");
    grafo K1 = crear_grafo();
    agregar_vertice(K1, et1);
    
    if(es_grafo_vacio(K1))
        printf("El grafo es vacio.\n");
    else
        printf("El grafo es NO vacio.\n");
    
    if(nro_vertices(K1) == 1)
        printf("El vertice fue agregado.\n");
    
    if(es_vertice(K1, et1))
        printf("El grafo TIENE el vertice %c. \n", et1);
    else
        printf("El grafo NO TIENE el vertice %c. \n", et1);
    if(es_vertice(K1, et2))
        printf("El grafo TIENE el vertice %c. \n", et2);
    else
        printf("El grafo NO TIENE el vertice %c. \n", et2);
    if(es_completo(K1))
        printf("El grafo ES COMPLETO.\n");
    else
        printf("El grafo NO ES COMPLETO.\n");
    eliminar_grafo(K1);
    printf("\n");

    // GRAFO K2
    printf("PRUEBA CON K2\n");
    grafo K2 = crear_grafo();
    agregar_vertice(K2, et1);
    agregar_vertice(K2, et2);
    printf("El grado del vertice %c es %d. \n", et1, grado(K2, et1));
    printf("El grado del vertice %c es %d. \n", et2, grado(K2, et2));
    num_t dist = distancia(K2, et1, et2);
    printf("La distancia entre el vertice %c y %c es %d. \n", et1, et2, dist);
    if(son_adyacentes(K2, et1, et2))
        printf("Los vertices son adyacentes.\n");
    else 
        printf("Los vertices NO son adyacentes.\n");
    if(es_completo(K2))
        printf("El grafo ES COMPLETO.\n");
    else
        printf("El grafo NO ES COMPLETO.\n");
    
    // luego de agregar la arista
    agregar_arista(K2, et1, et2);
    dist = distancia(K2, et1, et2);
    printf("El grado del vertice %c es %d. \n", et1, grado(K2, et1));
    printf("El grado del vertice %c es %d. \n", et2, grado(K2, et2));
    if(son_adyacentes(K2, et1, et2))
        printf("Los vertices son adyacentes. \n");
    else 
        printf("Los vertices NO son adyacentes.\n");
    printf("La distancia entre el vertice %c y %c es %d. \n", et1, et2, dist);
    if(es_completo(K2))
        printf("El grafo ES COMPLETO.\n");
    else
        printf("El grafo NO ES COMPLETO.\n");
    
    // luego de quitar la arista
    remover_arista(K2, et1, et2);
    dist = distancia(K2, et1, et2);
    printf("El grado del vertice %c es %d. \n", et1, grado(K2, et1));
    printf("El grado del vertice %c es %d. \n", et2, grado(K2, et2));
    printf("La distancia entre el vertice %c y %c es %d. \n", et1, et2, dist);
     if(son_adyacentes(K2, et1, et2))
        printf("Los vertices son adyacentes.\n");
    else 
        printf("Los vertices NO son adyacentes.\n");
    if(es_completo(K2))
        printf("El grafo ES COMPLETO.\n");
    else
        printf("El grafo NO ES COMPLETO.\n");
    if(es_arbol(K2))
        printf("El grafo ES ARBOL.\n");
    else
        printf("El grafo NO ES ARBOL.\n");
    eliminar_grafo(K2);
    printf("\n");

    // GRAFO K3
    printf("PRUEBA CON K3\n");
    grafo K3 = crear_grafo();
    // agregar 3 vertices
    agregar_vertice(K3, et1);
    agregar_vertice(K3, et2);
    agregar_vertice(K3, et3);
    // agregar 2 aristas
    agregar_arista(K3, et1, et2);
    agregar_arista(K3, et1, et3);

    if(es_aciclico(K3))
        printf("El grafo ES ACICLICO.\n");
    else
        printf("El grafo NO ES ACICLICO.\n");
    if(es_arbol(K3))
        printf("El grafo ES ARBOL.\n");
    else
        printf("El grafo NO ES ARBOL.\n");
    if(existe_rec_euleriano(K3))
        printf("El grafo TIENE RECORRIDO EULERIANO.\n");
    else
        printf("El grafo NO TIENE RECORRIDO EULERIANO.\n");
    if(existe_circ_euleriano(K3))
        printf("El grafo TIENE CIRCUITO EULERIANO.\n");
    else
        printf("El grafo NO TIENE CIRCUITO EULERIANO.\n");

    // agregar la ultima arista
    agregar_arista(K3, et3, et2);

    if(es_aciclico(K3))
        printf("El grafo ES ACICLICO.\n");
    else
        printf("El grafo NO ES ACICLICO.\n");
    if(es_arbol(K3))
        printf("El grafo ES ARBOL.\n");
    else
        printf("El grafo NO ES ARBOL.\n");
    if(existe_rec_euleriano(K3))
        printf("El grafo TIENE RECORRIDO EULERIANO.\n");
    else
        printf("El grafo NO TIENE RECORRIDO EULERIANO.\n");
    if(existe_circ_euleriano(K3))
        printf("El grafo TIENE CIRCUITO EULERIANO.\n");
    else
        printf("El grafo NO TIENE CIRCUITO EULERIANO.\n");

    eliminar_grafo(K3);
    printf("\n");

    // GRAFO K4
    printf("PRUEBA CON K4\n");
    grafo K4 = crear_grafo();
    // agregar 4 vertices
    agregar_vertice(K4, et1);
    agregar_vertice(K4, et2);
    agregar_vertice(K4, et3);
    agregar_vertice(K4, et4);

    if(es_conexo(K4))
        printf("El grafo ES CONEXO.\n");
    else
        printf("El grafo NO ES CONEXO.\n");
    printf("El grafo tiene %d componentes conexas. \n", componentes_conexas(K4));
    if(existe_rec_euleriano(K4))
        printf("El grafo TIENE RECORRIDO EULERIANO.\n");
    else
        printf("El grafo NO TIENE RECORRIDO EULERIANO.\n");
    if(existe_circ_euleriano(K4))
        printf("El grafo TIENE CIRCUITO EULERIANO.\n");
    else
        printf("El grafo NO TIENE CIRCUITO EULERIANO.\n");
    if(es_aciclico(K3))
        printf("El grafo ES ACICLICO.\n");
    else
        printf("El grafo NO ES ACICLICO.\n");

    // agregar 2 atistas
    agregar_arista(K4, et1, et2);
    agregar_arista(K4, et3, et4);    

    if(es_conexo(K4))
        printf("El grafo ES CONEXO.\n");
    else
        printf("El grafo NO ES CONEXO.\n");
    printf("El grafo tiene %d componentes conexas. \n", componentes_conexas(K4));    
    if(existe_rec_euleriano(K4))
        printf("El grafo TIENE RECORRIDO EULERIANO.\n");
    else
        printf("El grafo NO TIENE RECORRIDO EULERIANO.\n");
    if(existe_circ_euleriano(K4))
        printf("El grafo TIENE CIRCUITO EULERIANO.\n");
    else
        printf("El grafo NO TIENE CIRCUITO EULERIANO.\n");
    if(es_aciclico(K3))
        printf("El grafo ES ACICLICO.\n");
    else
        printf("El grafo NO ES ACICLICO.\n");

    // agregar el resto de las aristas (repitiendo una)
    agregar_arista(K4, et1, et3);
    agregar_arista(K4, et1, et4);
    // repite esta
    agregar_arista(K4, et2, et1);
    agregar_arista(K4, et2, et3);
    agregar_arista(K4, et2, et4);    
    
    if(es_conexo(K4))
        printf("El grafo ES CONEXO.\n");
    else
        printf("El grafo NO ES CONEXO.\n");
    printf("El grafo tiene %d componentes conexas. \n", componentes_conexas(K4));
    if(existe_rec_euleriano(K4))
        printf("El grafo TIENE RECORRIDO EULERIANO.\n");
    else
        printf("El grafo NO TIENE RECORRIDO EULERIANO.\n");
    if(existe_circ_euleriano(K4))
        printf("El grafo TIENE CIRCUITO EULERIANO.\n");
    else
        printf("El grafo NO TIENE CIRCUITO EULERIANO.\n");
    if(es_aciclico(K3))
        printf("El grafo ES ACICLICO.\n");
    else
        printf("El grafo NO ES ACICLICO.\n");

    eliminar_grafo(K4);
    printf("\n");
    // NOTA: HASTA ACA LA SALIDA ESTA BIEN
/*
    // GRAFO K5
    printf("PRUEBA CON K5\n");    
    grafo v5 = crear_grafo();
    // agregar 5 vertices
    agregar_vertice(v5, et1);
    agregar_vertice(v5, et2);
    agregar_vertice(v5, et3);
    agregar_vertice(v5, et4);
    agregar_vertice(v5, et5);
    // prueba de agregar uno ya existente
    agregar_vertice(v5, et1);

    eliminar_grafo(v5);

    // usar el complementario
    grafo K5 = complementario(v5);

    eliminar_grafo(K5);
    
    // GRAFO K3,3
    printf("PRUEBA CON K3,3\n");
    // grupo A = impares = {1, 3, 5}
    // grupo B = pares = {2, 4, 6}
    grafo K33 = crear_grafo();
    // agregar 6 vertices
    agregar_vertice(K33, et1);
    agregar_vertice(K33, et2);
    agregar_vertice(K33, et3);
    agregar_vertice(K33, et4);
    agregar_vertice(K33, et5);
    agregar_vertice(K33, et6);
    // aristas desde et1
    agregar_arista(K33, et1, et2);
    agregar_arista(K33, et1, et4);
    agregar_arista(K33, et1, et6);
    // aristas desde et2
    agregar_arista(K33, et2, et1);
    agregar_arista(K33, et2, et3);
    agregar_arista(K33, et2, et5);
    // aristas desde et3
    agregar_arista(K33, et3, et2);
    agregar_arista(K33, et3, et4);
    agregar_arista(K33, et3, et6);
    // aristas desde et4
    agregar_arista(K33, et4, et1);
    agregar_arista(K33, et4, et3);
    agregar_arista(K33, et4, et5);
    // aristas desde et5
    agregar_arista(K33, et5, et2);
    agregar_arista(K33, et5, et4);
    agregar_arista(K33, et5, et6);
    // aristas desde et6
    agregar_arista(K33, et6, et1);
    agregar_arista(K33, et6, et3);
    agregar_arista(K33, et6, et5);

    eliminar_grafo(K33);  

    // GRAFO DE PETERSEN
    grafo Petersen = crear_grafo();
    // agregar 10 vertices
    agregar_vertice(Petersen, et1);
    agregar_vertice(Petersen, et2);
    agregar_vertice(Petersen, et3);
    agregar_vertice(Petersen, et4);
    agregar_vertice(Petersen, et5);
    agregar_vertice(Petersen, et6);
    agregar_vertice(Petersen, et7);
    agregar_vertice(Petersen, et8);
    agregar_vertice(Petersen, et9);
    agregar_vertice(Petersen, et10);

    eliminar_grafo(Petersen);
*/
/*
    grafo C4 = crear_grafo();
    grafo P6 = crear_grafo();
*/

    return 0;
}
