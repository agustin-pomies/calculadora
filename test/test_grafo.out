PRUEBA CON GRAFO VACIO
El grafo es vacio.
El grafo NO TIENE el vertice A.

PRUEBA CON K1
El grafo es NO vacio.
El vertice fue agregado.
El grafo TIENE el vertice A. 
El grafo NO TIENE el vertice B. 
El grafo ES COMPLETO.

PRUEBA CON K2
El grado del vertice A es 0. 
El grado del vertice B es 0. 
La distancia entre el vertice A y B es -1. 
Los vertices NO son adyacentes.
El grafo NO ES COMPLETO.
El grado del vertice A es 1. 
El grado del vertice B es 1. 
Los vertices son adyacentes. 
La distancia entre el vertice A y B es 1. 
El grafo ES COMPLETO.
El grado del vertice A es 0. 
El grado del vertice B es 0. 
La distancia entre el vertice A y B es -1. 
Los vertices NO son adyacentes.
El grafo NO ES COMPLETO.
El grafo NO ES ARBOL.

PRUEBA CON K3
El grafo ES ARBOL.
El grafo TIENE RECORRIDO EULERIANO.
El grafo NO TIENE CIRCUITO EULERIANO.
El grafo NO ES ARBOL.
El grafo NO TIENE RECORRIDO EULERIANO.
El grafo TIENE CIRCUITO EULERIANO.

PRUEBA CON K4
El grafo NO ES CONEXO.
El grafo tiene 4 componentes conexas. 
El grafo NO TIENE RECORRIDO EULERIANO.
El grafo NO TIENE CIRCUITO EULERIANO.
El grafo NO ES CONEXO.
El grafo tiene 2 componentes conexas. 
El grafo NO TIENE RECORRIDO EULERIANO.
El grafo NO TIENE CIRCUITO EULERIANO.
El grafo ES CONEXO.
El grafo tiene 1 componentes conexas. 
El grafo NO TIENE RECORRIDO EULERIANO.
El grafo NO TIENE CIRCUITO EULERIANO.

